plugin_qa/PluginQualityTest.php

```diff
@@ -0,0 +1,490 @@
+    }
+
+    public function test_PluginDotJsonFile_HasCorrectFields_IfPluginIsClosedSource()
+    {
+        if ($this->getRepoOwner() != "piwikpro") {
+            return;
+        }
+
+        $this->assertArrayHasKey("homepage", $this->pluginJsonContents);
+        $this->assertEquals("https://piwik.pro/plugins", $this->pluginJsonContents["homepage"]);
+
+        $this->assertArrayHasKey("authors", $this->pluginJsonContents);
+        $this->assertEquals(array("name" => "Piwik PRO", "email" => "contact@piwik.pro", "homepage" => "https://piwik.pro"),
+            $this->pluginJsonContents["authors"]);
+
+        $this->assertArrayHasKey("license", $this->pluginJsonContents);
+        $this->assertEquals("Paid plugin", $this->pluginJsonContents["license"]);
+    }
+
+    public function test_PluginPhpFiles_HaveProperCommentHeader()
+    {
+        $expectedHeader = $this->getExpectedPluginHeader();
+        $this->assertPhpFilesHaveHeader($expectedHeader, $this->getPluginCodeFilesOfType('.php'));
+    }
+
+    public function test_PluginJsFiles_HaveProperCommentHeader()
+    {
+        $expectedHeader = $this->getExpectedPluginHeader();
+        $this->assertFilesHaveHeader($expectedHeader, $this->getPluginCodeFilesOfType('.js'));
+    }
+
+    public function test_PluginReadme_Exists()
+    {
+        $pluginReadmePath = $this->getPluginReadmePath();
+        $this->assertFileExists($pluginReadmePath);
+    }
+
+    /**
+     * @depends test_PluginReadme_Exists
+     * @dataProvider getPluginReadmeRequiredSections
+     */
+    public function test_PluginReadme_HasRequiredSection($sectionName)
+    {
+        $pluginReadme = $this->getPluginReadme();
+
+        $regex = '/^#+\s*' . preg_quote($sectionName) . '/im';
+        if (!preg_match($regex, $pluginReadme)) {
+            $this->fail("Plugin README.md is missing required section '$sectionName'.");
+        }
+    }
+
+    public function getPluginReadmeRequiredSections()
+    {
+        return array(
+            array('description'),
+            array('changelog'),
+            array('support'),
+        );
+    }
+
+    public function test_PluginReadme_HasCorrectTravisBuildBadges()
+    {
+        $pluginReadme = $this->getPluginReadme();
+        $this->assertHasBuildBadge('master', $pluginReadme);
+
+        if ($this->pluginUsesGitFlow) {
+            $this->assertHasBuildBadge('develop', $pluginReadme);
+        }
+    }
+
+    public function test_PluginReadme_HasCorrectSupportSection()
+    {
+        $pluginReadme = $this->getPluginReadme();
+
+        // get support section
+        if (!preg_match('/#+\s*Support(.*)(?=(?:\n#)|$)/is', $pluginReadme, $matches)) {
+            throw new \Exception("Cannot find Support section in plugin README.md.");
+        }
+
+        $supportSection = $matches[1];
+
+        if ($this->getRepoOwner() == 'piwikpro') {
+            $this->assertSupportSectionContainsPiwikProEmail($supportSection);
```

Comment:
To avoid such if statements it might be good to refactor code to have a PiwikQualityTest and PiwikProQuality test. This would make it better readable and kinda easier to maintain. They both could use a third class to not have much duplicated code. Also it would maybe make this third class testable. 

eg 
```php
class PiwikQualityTest {}
class PiwikProQualityTest {}
class QualityTest {
   public function getRepoOwner();
   public function getPluginRepoDescription();
   ....
}
```

This would make many private methods testable. Ideally there would be another `class ThirdPartyPluginQualityTest` that does some basic checks for third party developers so anyone benefits from it. Otherwise it would be not really cool to have such PiwikPro specific stuff in core